File: day13.js
import { readInput } from "./helpers/fileReader.js";

const input = readInput("./input/day13.txt");

const earliestTimestamp = parseInt(input[0]);
const runningBusses = input[1].split(",").filter((id) => {
  return id !== "x";
});

let nextEarliest = 0;
let counter = earliestTimestamp;
let busId = 0;
while (!nextEarliest) {
  for (let i = 0; i < runningBusses.length; i++) {
    if (counter % runningBusses[i] === 0) {
      nextEarliest = counter;
      busId = runningBusses[i];
      break;
    }
  }
  counter++;
}

console.log("Part 1:", busId * (nextEarliest - earliestTimestamp));

//find first large timestamp that is divisible by the first bus
let timestamp = 0;
let found = false;

let allBusses = input[1].split(",");
allBusses = allBusses.map(val => {
  return val === 'x' ? 1 : parseInt(val);
});

console.log('Part 2:',
  allBusses.slice(1).reduce(
    ([last, multiplier], current, i) => {
      for (let found = +last; ; found += multiplier) {
        if ((found + i + 1)%current === 0) {
          return [found, multiplier*current];
        }
      }
    },
    [allBusses[0], allBusses[0]]
  )[0]
);

File: 2020/day13.py
from functools import reduce
print('Python')
file = open('./input/day13.txt')
file = [x.strip() for x in file.readlines()]

earliestTimestamp = int(file[0])
runningBusses = list(filter(lambda id: id != 'x', file[1].split(',')))

nextEarliest = 0
counter = earliestTimestamp
busId = 0
while not nextEarliest:
    for i in range(len(runningBusses)):
        if counter % int(runningBusses[i]) == 0:
            nextEarliest = counter
            busId = int(runningBusses[i])
            break
    counter += 1

print('Part 1 : {}'.format(busId * (nextEarliest - earliestTimestamp)))

allBusses = file[1].split(',')
allBusses = list(map(lambda val: 1 if val == 'x' else int(val), allBusses))


def findBus(last, multiplier, current, i):
    found = last
    last += 1
    while True:
        if found + i + 1 % current == 0:
            return [found, multipier * current]
        found += multiplier


print('Part 2 : {}'.format(reduce(findBus(), [allBusses[0], allBusses[0]])))
